standardtracer/tracer.go

```diff
@@ -2,16 +2,67 @@ package standardtracer
 
 import (
 	"sync"
+	"sync/atomic"
 	"time"
 
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
-// New creates and returns a standard Tracer which defers to `recorder` after
-// RawSpans have been assembled.
-func New(recorder SpanRecorder) opentracing.Tracer {
+type sampleFunc func(int64) bool
+
+// Options allows creating a customized Tracer via NewWithOptions. The object
+// is thread safe and can be updated freely to influence the behavior of the
+// Tracer.
+type Options struct {
+	shouldSample       atomic.Value // contains a sampleFunc
+	trimUnsampledSpans int32        // updated atomically
+	recorder           atomic.Value // contains a SpanRecorder
```

Comment:
@tschottdorf why can't the sampleFunc handle any realtime adjustments internally?

My concern here is not performance, but just plain-old complexity. We have taken what would be a straightforward Options struct and turned it into something with subtlety and opacity.

Also, FWIW, it's important to record the sampling rate somewhere if statistical statements are to be made about the recorded traces later. That's why "dapperish" uses more than a bool to record the sampling state (dapper recorded the sampling *decision* as a bit, but also packed the reciprocal sampling rate in there as well). Finally, a truly smart sampler would want more than just the trace_id to make the decision, but that seems out of scope.

However, stepping way back from the minutia of this change: it worries me that we're debating things this impl-specific in any `opentracing-*` repository. It seems like `standardtracer` is quickly becoming `cockroachtracer` :)  Which is not necessarily a bad thing, as it was just a toy before. But a lot of the details of the "standard" API are making assumptions about implementation priorities that may not generalize well.